crates/namada/src/ledger/protocol/mod.rs

```diff
@@ -432,166 +482,168 @@ where
-    S: State + StorageRead + StorageWrite,
+    S: State<D = D, H = H>
+        + StorageRead
+        + Read<Err = namada_state::Error>
+        + Sync,
+    D: 'static + DB + for<'iter> DBIter<'iter> + Sync,
+    H: 'static + StorageHasher + Sync,
+    CA: 'static + WasmCacheAccess + Sync,
 {
-    let balance = crate::token::read_balance(
-        state,
-        &wrapper.fee.token,
-        &wrapper.fee_payer(),
-    )
-    .unwrap();
-
-    const FEE_PAYMENT_DESCRIPTOR: std::borrow::Cow<'static, str> =
-        std::borrow::Cow::Borrowed("wrapper-fee-payment");
-
     match wrapper.get_tx_fee() {
         Ok(fees) => {
-            let fees =
-                crate::token::denom_to_amount(fees, &wrapper.fee.token, state)
-                    .map_err(|e| Error::FeeError(e.to_string()))?;
+            let fees = crate::token::denom_to_amount(
+                fees,
+                &wrapper.fee.token,
+                shell_params.state,
+            )
+            .map_err(Error::StorageError)?;
 
-            let current_block_height =
-                state.in_mem().get_last_block_height().next_height();
+            let balance = crate::token::read_balance(
+                shell_params.state,
+                &wrapper.fee.token,
+                &wrapper.fee_payer(),
+            )
+            .map_err(Error::StorageError)?;
 
-            if let Some(post_bal) = balance.checked_sub(fees) {
-                token_transfer(
-                    state,
+            let (post_bal, valid_batched_tx_result) = if let Some(post_bal) =
+                balance.checked_sub(fees)
+            {
+                fee_token_transfer(
+                    shell_params.state,
                     &wrapper.fee.token,
                     &wrapper.fee_payer(),
                     block_proposer,
                     fees,
-                )
-                .map_err(|e| Error::FeeError(e.to_string()))?;
+                )?;
 
-                let target_post_balance = Some(
-                    namada_token::read_balance(
-                        state,
+                (Some(post_bal), None)
+            } else {
+                // See if the first inner transaction of the batch pays the fees
+                // with a masp unshield
+                if let Ok(Some(valid_batched_tx_result)) =
+                    try_masp_fee_payment(shell_params, tx, tx_index)
+                {
+                    // NOTE: Even if the unshielding was succesfull we could
+                    // still fail in the transfer (e.g. cause the unshielded
+                    // amount is not enough to cover the fees). In this case we
+                    // want do drop the changes applied by the masp transaction
+                    // and try to drain the fees from the transparent balance.
+                    // Because of this we must NOT propagate errors from within
+                    // this branch
+                    let balance = crate::token::read_balance(
```

Comment:
just out of curiosity, what is the flow where this is expected? My surface level understanding seems to think we should just reject the tx at this point.

Reply:
In theory this should never happen because a tx that cannot pay fees cannot be included in a block (we reject `process_proposal`). Here we have this logic to guarantee liveness and some degree of correcteness in case the check in process

Reply:
Not sure that I follow - why would we want to subtract from a transparent balance, if the user tried to pay fees by unshielding? Sounds confusing. I feel like we should just bail if the unshielding doesn't unshield enough.

Reply:
It's just a fallback in case the check in `process_proposal` was flawed. We shouldn't really ever reach this branch but, in case, the logic here was to try and get at least a partial fee payment from the transparent balance of the wrapper's signer. We can also decide to drop this logic and just do nothing. Of course, the tx is rejected

Reply:
Removed it in e1b5857845